### ISSUE TITLE
Secured Push Notifications not loading content on ios

### ISSUE THREAD

**Issue by james-surevine** [RocketChat/Rocket.Chat.ReactNative#4199]
### Description:

Having installed a trial version of Rocket.Chat enterprise in order to test the 'Secured Push Notifications' feature. It appears this does not work as expected. My android client works well and the notification loads the content (by message ID) as expected. For ios however, the notification content does not load and the notification shows a generic 'You have a new message' message.

Debugging shows that 
1) the call from the ios app to /api/v1/push.get?id=... gets a 401. It tries 5 times to load the notification but then gives up and displays the generic message. Further investigation shows that the ios app does not include a `x-auth-token` header.
2) The call from the android app to /api/v1/push.get?id= does include a `x-auth-token` header and returns successfully with a HTTP 200 response

### Environment Information:

- Rocket.Chat Server Version: 4.7.0 
- Rocket.Chat App Version: 4.27.1 (ios app)
- Device Name: Iphone 12
- OS Version: 15.4.1

### Steps to reproduce:

1. Installed Rocket.Chat using docker-compose
2. Registered the installation and enabled a trial version of Enterprise Edition
4. Create admin user user1
3. Create user2 on browser client
4. Log in to ios app as user2 and close the app
5. Log in to browser client as user1
6. send a message in the browser client as user1, tagging user 2, e.g. `@user2 test1`
7. Observe that after a delay a Push notification is received in the ios app, however only a generic 'you have a new message'

### Expected behavior:
ios should display a push notification with the content 'test1'

### Actual behavior:
ios (After a delay) displays a push notification with generic content of 'you have a new message'

### Additional context:
as mentioned above, use tcpdump and analyse to show that the call to load the notification by ID from the ios app did not include a `x-auth-token` and resulted in a 401 (Unauthorized). This happens 5 times before the app displays the generic notification text.

**Comment by ankar84** [comment]
@james-surevine agree with Diego, great investigation!

With Secure Push feature a have a bit different experience.
Most of mobile clients are works perfect, they show message content in Secured Push.
But there is some clients (a lot of them actually) that shows generic message.
Majority of clients with that issue is definitely iOS clients.
But I have users reports about same generic message on Android too!
Here is a screen:
![image](https://user-images.githubusercontent.com/4023037/168724710-57818eda-d662-4ade-a010-993b25a0b45e.png)
Here is an iOS screen
![image](https://user-images.githubusercontent.com/4023037/168724810-3b0a65cb-5bb3-4491-8106-2d4b09868a2d.png)

There is a temporary workaround - delete app and then install it again. That manipulation fixes Secured Push for some time, but not permanently - it will broke again.
And I definitely saw when user first gets generic push message, but then after some time and without re-install of mobile application that same user gets proper message content.

So, my guess is not the only mobile client issue, it could be some server related issue too, or even both.
@sampaiodiego [agreed](https://github.com/RocketChat/Rocket.Chat/pull/25347#issuecomment-1118643332) with me.

In fact that wonderful feature should be polished to bright like a diamond.

**Comment by diegolmello** [comment]
@ankar84 I was not aware about that commit on server.
Can you keep us updated once it's landed and tested by your team?

**Comment by ankar84** [comment]
4.8.1 didn't fix that issue
![image](https://user-images.githubusercontent.com/4023037/173172385-ea66749e-b084-4e17-a1a9-28be0abb3efa.png)

**Comment by thakilla** [comment]
Well it does make no sense, but disabling the option "Fetch full message content from the server on receipt" under "Push -> Privacy" fixed all our problems while still getting all of the message content.

![Bildschirmfoto 2022-08-25 um 16 41 36](https://user-images.githubusercontent.com/449411/186696351-87b211b8-972f-47f6-a034-86809580df29.png)

**Comment by ankar84** [comment]
> Well it does make no sense, but disabling the option "Fetch full message content from the server on receipt" under "Push -> Privacy" fixed all our problems while still getting all of the message content.

It makes a lot of sense, because now all content of push message go through Apple and Google servers if `Fetch full message content from the server on receipt` is disabled.

**Comment by diegolmello** [comment]
@ankar84 @james-surevine I've been trying to reproduce this on Mobile 4.29 (and on the unreleased 4.30) and Server 5.0, but it's working all the time.
Going through logs I can see the requests are fetching the message content as expected.
Please run your tests and let me know if it's working for you.
We can reopen if needed.

**Comment by diegolmello** [comment]
Nevermind.
For some reason, it's started to fail.
I'm still investigating.

**Comment by ankar84** [comment]
As I told you here,  majority of the clients works perfect for some time, but then suddenly some of that clients start get dummy push messages. App reinstall fix that problem immediately, bit not permanently.

**Comment by wreiske** [comment]
This is affecting our instance as well after enabling the enterprise license.

**Comment by diegolmello** [comment]
We investigated this issue on https://github.com/RocketChat/Rocket.Chat.ReactNative/pull/4487 and we couldn't find anything related to iOS specifically, but we found it on Android.
We're aware it does happen on iOS though, so we're rolling out the Android only fix on next update and understand if it could be something on backend side.

**Comment by diegolmello** [comment]
@ankar84 You're right. I've been testing secured push on the last days and it's working all the time.
Let's make the release and you can try it on your side.

**Comment by ansiwen** [comment]
I have the same problem with the `message-id-only` notifications, but _only_ for public channels (type "c"), for private channels (type "p") it works as expected. I can see the ’/api/v1/push.get?id=...` requests in both cases and they are all answered with a `200 OK` from the RC server, with basically the same JSON data, but in the case of type "c" channels the requests come in many times from the App, until it gives up and just shows the generic notification ("you have a new message"). So it seems like the app ignores or rejects the type "c" notifications and tries to get them again and again.

**Comment by diegolmello** [comment]
@ansiwen Do you have E2EE enabled?

**Comment by ansiwen** [comment]
> @ansiwen Do you have E2EE enabled?

@diegolmello No. And I would be surprised if this works for anyone. The model ist simply not up to date:
https://github.com/RocketChat/Rocket.Chat.ReactNative/blob/c5e1b9e7fd5fed956e595babacdb9a7658b55867/ios/Shared/Models/RoomType.swift#L11-L14

**Comment by diegolmello** [comment]
@ansiwen 
That's not necessarily the issue.
It's just to append the username to the push content: `User sent a message` vs `Sent a message`.

Also you said `direct` isn't working as well.

**Comment by ansiwen** [comment]
> Also you said `direct` isn't working as well.

That's true. I assumed the JSON parsing fails and I have no idea, what else it could be, because the responses from the server look identical besides the `type` field. I would expect it must be something during this code: 

https://github.com/RocketChat/Rocket.Chat.ReactNative/blob/fd210c4713f7101ad91a627d516ee70dda7c9917/ios/NotificationService/NotificationService.swift#L28-L34

**Comment by ankar84** [comment]
> @ankar84 is this different behavior depending on the room type something that you can reproduce?

We just test case described by @ansiwen [here](https://github.com/RocketChat/Rocket.Chat.ReactNative/issues/4199#issuecomment-1500696939) and I can confirm that we see problem with generic push messages from public channels (type "c"). 
But direct message push in our test was ok.
So, guess this is a very good details to het root cause of that problem.

**Comment by ankar84** [comment]
Here is a recent experiment
Android
![image](https://user-images.githubusercontent.com/4023037/231099529-b18a6cf2-56e9-4e37-b848-2c2470ac3c2f.png)
iOS
![image](https://user-images.githubusercontent.com/4023037/231099790-f26dfab6-235c-4922-9190-45bfa480931f.png)
![image](https://user-images.githubusercontent.com/4023037/231099942-a04e93a7-ae95-4a72-8164-e3aad4b1876f.png)
![image](https://user-images.githubusercontent.com/4023037/231100095-31bb703b-f45d-489a-8078-e746f9a7f536.png)

@diegolmello Conclusion: **it's iOS only issue with public channels only**

**Comment by diegolmello** [comment]
Awesome. Thanks, Anton!

**Comment by ansiwen** [comment]
That all confirms the theory, that the unmarhaling of the JSON data is failing because of the outdated model, and I simply made a mistake with the direct messages.

**Comment by diegolmello** [comment]
Ok, I'm going to give it a try then.

**Comment by diegolmello** [comment]
@ansiwen It actually fixed it!
We're going to make a release later this week.
Thanks for your help!

**Comment by diegolmello** [comment]
@ankar84 Can you enable secured push on the next release?

**Comment by ankar84** [comment]
@diegolmello We are on 6.2.10 server version finally!
And base mobile version for our users is 4.39.0
We enabled Secured Push on our server today and definitely can tell that [this bug](https://github.com/RocketChat/Rocket.Chat.ReactNative/issues/4199#issuecomment-1502894369) is totally fixed!
But I see some reports, that generic "You have new message" still appears.
Still checking details with @shiryov

**Comment by ankar84** [comment]
Here is a one example of generic message received today
It's android, RC mobile version 4.39.0
![image](https://github.com/RocketChat/Rocket.Chat.ReactNative/assets/4023037/d94f7d37-86b5-43d1-92d5-dfd44dd80a74)

`push.get` endpoint look works good
![image](https://github.com/RocketChat/Rocket.Chat.ReactNative/assets/4023037/a168b687-7396-4211-b4f3-3aa4e4fe32fd)

**Comment by shiryov** [comment]
Hello @diegolmello I'm from @ankar84 team 
yes, we have a rate limiter enabled with these parameters:
![image](https://github.com/RocketChat/Rocket.Chat.ReactNative/assets/666827/2ab2a2c5-d2b6-42b4-bf2b-900a97a782c9)

Maybe a problem in the long execution of the request, 2 seconds is not long?
![image](https://github.com/RocketChat/Rocket.Chat.ReactNative/assets/666827/7dc1c67a-4466-486f-985c-dc85ab5a215e)

**Comment by diegolmello** [comment]
If you changed it to 200, it might not be a rate limiter issue.
Can you disable rate limiter completely just to test this?
It shouldn't affect your operation that much for a quick test.
(I could argue 200 is almost like the rate limiter is disabled already...)

> Maybe a problem in the long execution of the request, 2 seconds is not too long for android/ios push service?

I don't think so. We have 30 seconds to fetch and change the content of a push notification once it's received.
2 seconds for a request is not unusual.

**Comment by shiryov** [comment]
Hello @diegolmello 
Completely disabling rate limiter seems to have solved the problem. 
Already 3 working days (with disabled limiter) security pushes come to both Android and Apple devices with text. 
But on product environment rate limiter is required.
With a limit of 200 per minute, pushes definitely does not work. Based on our graph and the number of instances - 20, in order for push notifications to work, we need to increase the limit to 50000/20=**2500**, right?

![image](https://github.com/RocketChat/Rocket.Chat.ReactNative/assets/666827/e050221f-36cc-4c4a-8bba-cf1c828adc5a)

**Comment by diegolmello** [comment]
Glad it worked!
I'm not an expert on our rate limiting system.
Perhaps you can move this discussion to a place where backenders can see it?

**Comment by pierre-lehnen-rc** [comment]
You should not be hitting the rate limiter with those numbers, specially not very frequently. Is the issue still present on both platforms, or mostly one over the other?
Do you have any metric with number of push notifications sent per minute that you could compare to the number of requests to push.get?

One possibility I can think of is that the rate limiter is for some reason counting requests from multiple/all users as if they were from the same origin, then hitting the limit much earlier because of that. Is there anything out of ordinary in your network that could maybe affect this?
